amqp/types.py

```diff
@@ -0,0 +1,566 @@
+    def delivery_tag(self) -> str:
+        ...
+
+
+class AbstractChannelT(metaclass=abc.ABCMeta):
+    """Abstract channel type."""
+
+    connection: 'ConnectionT'
+    channel_id: int
+    auto_decode: bool = False
+    is_open: bool = False
+
+    @abc.abstractmethod
+    def __enter__(self) -> 'AbstractChannelT':
+        ...
+
+    @abc.abstractmethod
+    def __exit__(self, *exc_info) -> None:
+        ...
+
+    @abc.abstractmethod
+    async def __aenter__(self) -> 'AbstractChannelT':
+        ...
+
+    @abc.abstractmethod
+    async def __aexit__(self, *exc_info) -> None:
+        ...
+
+    @abc.abstractmethod
+    def _setup_listeners(self):
+        ...
+
+    @abc.abstractmethod
+    async def send_method(
+            self, sig: method_sig_t,
+            format: str = None,
+            args: Sequence = None,
+            *,
+            content: MessageT = None,
+            wait: WaitMethodT = None,
+            callback: Callable = None,
+            returns_tuple: bool = False) -> Thenable:
+        ...
+
+    @abc.abstractmethod
+    async def close(
+            self,
+            *,
+            reply_code: int = 0,
+            reply_text: str = '',
+            method_sig: method_sig_t = method_sig_t(0, 0),
+            argsig: str = 'BsBB') -> None:
+        ...
+
+    @abc.abstractmethod
+    def collect(self) -> None:
+        ...
+
+    @abc.abstractmethod
+    async def wait(
+            self,
+            method: WaitMethodT,
+            *,
+            callback: Callable = None,
+            timeout: float = None,
+            returns_tuple: bool = False) -> Any:
+        ...
+
+    async def dispatch_method(
+            self,
+            method_sig: method_sig_t,
+            payload: bytes,
+            content: MessageT) -> None:
+        ...
+
+
+class ConnectionT(AbstractChannelT):
+    """Connection channel type."""
+
+    Channel: type
+    Transport: type
+
+    host: str
+    userid: str
+    password: str
+    login_method: str
+    login_response: Any
+    virtual_host: str
+    locale: str
+    client_properties: MutableMapping
+    ssl: Any
+    channel_max: int
+    frame_max: int
+    on_open: Thenable
+    on_tune_ok: Thenable
+    confirm_publish: bool
+    connect_timeout: float
+    read_timeout: float
+    write_timeout: float
+    socket_settings: Mapping
+
+    negotiate_capabilities: Mapping[str, bool]
+    library_properties: Mapping[str, Any]
+    heartbeat: float
+    client_heartbeat: float
+    server_heartbeat: float
+    last_heartbeat_sent: float
+    last_heartbeat_received: float
+    bytes_sent: int = 0
+    bytes_recv: int = 0
+    prev_sent: int
+    prev_recv: int
+
+    connection_errors: Tuple[type, ...]
+    channel_errors: Tuple[type, ...]
+    recoverable_connection_errors: Tuple[type, ...]
+    recoverable_channel_errors: Tuple[type, ...]
+
+    transport: TransportT
+    channels: MutableMapping[int, AbstractChannelT]
+    loop: asyncio.AbstractEventLoop
```

Comment:
anyio/trio alternatives?

Reply:
There is no alternative to the loop object in trio.
You should familiarize yourself with Task Groups (aka Nurseries in trio's terminology).